doc/api/webcrypto.md

```diff
@@ -0,0 +1,1672 @@
+
+async function generateAesKey(length = 256) {
+  const key = await subtle.generateKey({
+    name: 'AES-CBC',
+    length
+  }, true, ['encrypt', 'decrypt']);
+
+  return key;
+}
+```
+
+#### Elliptic curve key pairs
+
+```js
+const { subtle } = require('crypto').webcrypto;
+
+async function generateEcKey(namedCurve = 'P-521') {
+  const {
+    publicKey,
+    privateKey
+  } = await subtle.generateKey({
+    name: 'ECDSA',
+    namedCurve,
+  }, true, ['sign', 'verify']);
+
+  return { publicKey, privateKey };
+}
+```
+
+#### HMAC keys
+
+```js
+const { subtle } = require('crypto').webcrypto;
+
+async function generateHmacKey(hash = 'SHA-256') {
+  const key = await subtle.generateKey({
+    name: 'HMAC',
+    hash
+  }, true, ['sign', 'verify']);
+
+  return key;
+}
+```
+
+#### RSA key pairs
+
+```js
+const { subtle } = require('crypto').webcrypto;
+const publicExponent = new Uint8Array([1, 0, 1]);
+
+async function generateRsaKey(modulusLength = 2048, hash = 'SHA-256') {
+  const {
+    publicKey,
+    privateKey
+  } = await subtle.generateKey({
+    name: 'RSASSA-PKCS1-v1_5',
+    modulusLength,
+    publicExponent,
+    hash,
+  }, true, ['sign', 'verify']);
+
+  return { publicKey, privateKey };
+}
+```
+
+### Encryption and decryption
+
+```js
+const { subtle, getRandomValues } = require('crypto').webcrypto;
+
+async function aesEncrypt(plaintext) {
+  const ec = new TextEncoder();
+  const key = await generateAesKey();
+  const iv = getRandomValues(new Uint8Array(16));
+
+  const ciphertext = await subtle.encrypt({
+    name: 'AES-CBC',
+    iv,
+  }, key, ec.encode(plaintext));
+
+  return {
+    key,
+    iv,
+    ciphertext
+  };
+}
+
+async function aesDecrypt(ciphertext, key, iv) {
+  const dec = new TextDecoder();
+  const plaintext = await subtle.decrypt({
+    name: 'AES-CBC',
+    iv,
+  }, key, ciphertext);
+
+  return dec.decode(plaintext);
+}
+```
+
+### Exporting and importing keys
+
+```js
+const { subtle } = require('crypto').webcrypto;
+
+async function generateAndExportHmacKey(format = 'jwk', hash = 'SHA-512') {
+  const key = await subtle.generateKey({
+    name: 'HMAC',
+    hash
+  }, true, ['sign', 'verify']);
+
+  return subtle.exportKey(format, key);
+}
+
+async function importHmacKey(keyData, format = 'jwk', hash = 'SHA-512') {
+  const key = await subtle.importKey(format, keyData, {
+    name: 'HMAC',
+    hash
+  }, true, ['sign', 'verify']);
+
+  return key;
+}
+```
+
+### Wrapping and unwrapping keys
+
+```js
+const { subtle } = require('crypto').webcrypto;
+
+async function generateAndWrapHmacKey(format = 'jwk', hash = 'SHA-512') {
+  const [
+    key,
+    wrappingKey
+  ] = await Promise.all([
+    subtle.generateKey({
+      name: 'HMAC', hash
+    }, true, ['sign', 'verify']),
+    subtle.generateKey({
+      name: 'AES-KW',
+      length: 256
+    }, true, ['wrapKey', 'unwrapKey'])
+  ]);
+
+  const wrappedKey = await subtle.wrapKey(format, key, wrappingKey, 'AES-KW');
+
+  return wrappedKey;
+}
```

Comment:
> [Error: Cipher job failed.]

Not sure if this is a sample or an implementation error.

Reply:
The issue here is that the JSON stringified output length is not a multiple of 8. We should be able to fix it by padding the output.

Reply:
More on this... Web Crypto allows JWK to be wrapped but it needs to be stringified and converted to UTF8. After doing so, I fixed this by adding padding white space to the end of the stringified json to ensure that the length is a multiple of 8.